### PR TITLE
RSDK-5187: 2d obstacles module

### PR DESCRIPTION
**Pull request by Rob1in**
# Algorithm

1. Gets PCD bytes from planar lidar
2. Decodes it into PointCloud
3. Finds clusters using DBSCAN
4. If there are some big clusters, refines with 2d Ransac
5. Find convex hull for each clusters using Graham Scan
6. For each convex hull, find minimum fitting bounding box using rotating calipers
7. Transform each bounding box into GeometriesInFrame
8. Encode the cluster (=planar point cloud) into PCD and returns it along side the GeometriesInFrame as a `PointCloudObject`

# Manual Test
On a Mac:
 configured like [this](https://github.com/Rob1in/obstacles-2d-lidar/tree/module#example-config) and called `get_object_point_clouds()` from a Python SDK. 
 
 will do RPi tomorrow.
# Important notes

1. As all the cameras are not accessible yet, the only camera that can be reached is the one in put in the attribute `camera_name` (it means that to call `get_object_point_clouds()`, the argument `camera_name` is not used **for the moment**).
2. PCD can only be decoded in binary (ok for RPLidar A1) and encoded in ascii.
3. [test_local.py](https://github.com/Rob1in/obstacles-2d-lidar/blob/module/src/test_local.py) might be useful for review

 # Open questions

1. orientation vector for the Geometry...
2. @bhaney ~I stole the `run.sh` that you put in visual odometry but haven't tested it well with my ugly `requirements.txt`. Will do at the same time as [this](https://viam.atlassian.net/browse/RSDK-5190) if it's fine?~ I just added [.env, Makefile, meta.json, run.sh, setup.sh](https://github.com/viamrobotics/obstacles-2d-lidar/pull/1/commits/d3dcb392a983a0c123a419f5933e8f19235e3f8b) and will do the github/workflows for release and upload to registry next

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,103 @@
...
+```
+
+## Config
+### Example config 
+```json
+{
+  "services": [
+    {
+      "namespace": "rdk",
+      "model": "viam:vision:obstacles-2d-detector",
```
Comment (bhaney) [inline README.md]: The model name specified in the config is "obstacles-2d-lidar". I understand that it's not necessary that the point clouds come from a lidar, but detector is a misnomer too, as this does not implement GetDetections. Consider just renaming it to the the scope's obstacles-2d-lidar, or if you feel strongly that the name should be different, we can propose and update the scope to be obstacles-2d-pointcloud or something similar

File: meta.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "viam:obstacles-2d-detetctor",
```
Comment (bhaney) [inline meta.json]: change to obstacles-2d-lidar

File: meta.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  "visibility": "public",
+  "url": "https://github.com/viamrobotics/obstacles-2d-lidar",
+  "description": "This vision service uses planar lidar as input to detect obstacles. It implements the method get_object_point_poin_clouds() of the vision service API.",
+  "models": [
+    {
+      "api": "rdk:service:vision",
+      "model": "viam:vision:obstacles-2d-detetctor"
```
Comment (bhaney) [inline meta.json]: change to obstacles-2d-lidar

File: src/obstacles_2d_module.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+from viam.logging import getLogger
+from detector import Detector
+import pointcloud.point_cloud as pointcloud
+from pointcloud.decode_pcd import *
+from pointcloud.encode_pcd import *
+
+LOGGER = getLogger(__name__)
+
+class ObstacleDetectorModule(Vision, Reconfigurable):
+    MODEL: ClassVar[Model] = Model(ModelFamily("viam", "vision"), "obstacles-2d-detector")
```
Comment (bhaney) [inline src/obstacles_2d_module.py]: change to obstacles-2d-lidar

File: src/graham_scan.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        y += y[:1]
+
+        # Plot the polygon
+        plt.style.use("dark_background")
+        plt.plot(x, y, marker=None, linestyle='dashed')
+
+        # Add labels and title
+        # plt.xlabel('X-axis')
+        # plt.ylabel('Y-axis')
+        # plt.title('Polygon Plot')
```
Comment (bhaney) [inline src/graham_scan.py]: delete un-used code

File: src/obstacles_2d_module.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        service.reconfigure(config, dependencies)
+        return service
+    
+     # Validates JSON Configuration
+    @classmethod
+    def validate_config(cls, config: ServiceConfig) -> Sequence[str]:
+        camera_name = config.attributes.fields["camera_name"].string_value
+        if camera_name == "":
+            raise Exception(
+                "A camera name is required for 2d_obstacles_vision service.")
```
Comment (bhaney) [inline src/obstacles_2d_module.py]: change the name to its actual name, obstacles-2d-lidar

File: src/pointcloud/decode_pcd.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            metadata['HEIGHT'] = int(line.split(' ')[1])
+        elif line.startswith('VIEWPOINT'):
+            metadata['VIEWPOINT'] = [int(count) for count in line.split(' ')[1:]]
+        elif line.startswith('POINTS'):
+            metadata['POINTS'] = int(line.split(' ')[1])
+        elif line.startswith('DATA'):
+            metadata['DATA'] = line.split(' ')[1]
+
+    # Find the start of binary data, the following won't work if DATA ascii
+    data_start = header_lines.index(b'DATA binary') + 1
```
Comment (bhaney) [inline src/pointcloud/decode_pcd.py]: Can you put a check here and return a  clear error if the input pointcloud is not binary

File: src/pointcloud/encode_pcd.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
+from .point_cloud import PointCloud
+from typing import List
+
+class Encoder():
```
Comment (bhaney) [inline src/pointcloud/encode_pcd.py:4]: note: probably won't be necessary, but binary encoding is also a potential speed-up if that becomes necessary

File: src/pointcloud/point_cloud.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+# from .encode_pcd import *
+
+
+class PlanarPointCloud():
+    '''
+    self.X array of shape (n_points, 1)
+    self.Y array of shape (n_points, 1)
+    '''
+    def __init__(self, points:list=[], normalize:bool = False) -> None:
+        # self.points = np.array(points)
```
Comment (bhaney) [inline src/pointcloud/point_cloud.py]: delete line

File: src/pointcloud/point_cloud.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    def Y_norm(self):
+        if self.scale is None:
+            raise AttributeError("Planar poincloud must be normalized first")
+        return self.Y/self.scale 
+    
+    # @property
+    # def points_norm(self):
+    #     if self.scale is None:
+    #         raise AttributeError("Planar poincloud must be normalized first")
+    #     return self.points_norm
```
Comment (bhaney) [inline src/pointcloud/point_cloud.py]: delete un-used code

File: src/pointcloud/point_cloud.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    def Y(self):
+        ''' 
+        self.Y array of shape (n_points, 1)
+        '''
+        Y = self.points[:,1]
+        return Y.reshape(-1, 1)
+       
+    @property 
+    def X_norm(self):
+        if self.scale is None:
```
Comment (bhaney) [inline src/pointcloud/point_cloud.py:44]: You can normalize and cache the result if someone calls this, rather than creating an exception

**Review by bhaney — CHANGES_REQUESTED** [review]
Nice work, lots of cool implementations here, which can be nicely grown into something more sophisticated if needed. The biggest change I'm asking for is the name of the model - the scope doc required it to be "obstacles_2d_lidar"

File: README.md (hunk 1 above)
Comment (Rob1in) [reply README.md]: Sorry I don't know why I put that. The initial name seems very fine to me, changing to this one now

File: src/pointcloud/encode_pcd.py (hunk 8 above)
Comment (Rob1in) [reply src/pointcloud/encode_pcd.py:4]: PCD can even be 'binary_compressed' !

File: src/pointcloud/point_cloud.py (hunk 11 above)
Comment (Rob1in) [reply src/pointcloud/point_cloud.py:44]: done

File: src/pointcloud/point_cloud.py (hunk 10 above)
Comment (Rob1in) [reply src/pointcloud/point_cloud.py]: done

File: src/pointcloud/point_cloud.py (hunk 9 above)
Comment (Rob1in) [reply src/pointcloud/point_cloud.py]: done

File: src/pointcloud/decode_pcd.py (hunk 7 above)
Comment (Rob1in) [reply src/pointcloud/decode_pcd.py]: done

File: src/graham_scan.py (hunk 5 above)
Comment (Rob1in) [reply src/graham_scan.py]: done

File: src/obstacles_2d_module.py (hunk 4 above)
Comment (Rob1in) [reply src/obstacles_2d_module.py]: done

File: meta.json (hunk 3 above)
Comment (Rob1in) [reply meta.json]: done

File: meta.json (hunk 2 above)
Comment (Rob1in) [reply meta.json]: done

File: README.md (hunk 1 above)
Comment (Rob1in) [reply README.md]: done

**Comment by Rob1in** [comment]
Also added some changes in the import so it works with venv

**Review by bhaney — APPROVED** [review]
LGTM!

**Pull request merged**